c++/src/kj/compat/http.c++

```diff
@@ -2594,12 +2663,167 @@ private:
+              Z_DEFAULT_COMPRESSION,
+              Z_DEFLATED,
+              windowBits,
+              8,  // memLevel = 8 is the default
+              Z_DEFAULT_STRATEGY);
+          break;
+        case Mode::DECOMPRESS:
+          windowBits = -config.inboundMaxWindowBits.orDefault(15);
+          // We use negative values because we want to use raw inflate.
+          result = inflateInit2(&ctx, windowBits);
+          break;
+        default:
+          KJ_UNREACHABLE;
+      }
+      KJ_REQUIRE(result == Z_OK, "Failed to initialize compression context.");
+    }
+
+    ~Context() noexcept(false) {
+      switch (mode) {
+        case Mode::COMPRESS:
+          deflateEnd(&ctx);
+          break;
+        case Mode::DECOMPRESS:
+          inflateEnd(&ctx);
+          break;
+      }
+    }
+
+    KJ_DISALLOW_COPY(Context);
+
+    void setInput(const void* in, size_t size) {
+      ctx.next_in = const_cast<byte*>(reinterpret_cast<const byte*>(in));
+      ctx.avail_in = size;
+    }
+
+    Result pumpOnce() {
+      ctx.next_out = buffer;
+      ctx.avail_out = sizeof(buffer);
+
+      int result = Z_OK;
+
+      switch (mode) {
+        case Mode::COMPRESS:
+          result = deflate(&ctx, Z_SYNC_FLUSH);
+          KJ_REQUIRE(result == Z_OK || result == Z_BUF_ERROR || result == Z_STREAM_END,
+                      "Compression failed.");
+          break;
+        case Mode::DECOMPRESS:
+          result = inflate(&ctx, Z_SYNC_FLUSH);
+          KJ_REQUIRE(result == Z_OK || result == Z_BUF_ERROR || result == Z_STREAM_END,
+                      "Decompression failed.");
+          break;
+        default:
+          KJ_UNREACHABLE;
+      }
+
+      return Result {
+        .processResult = result,
+        .buffer = kj::arrayPtr(buffer, sizeof(buffer) - ctx.avail_out),
+      };
+    }
+
+    kj::OneOf<kj::Vector<kj::byte>, kj::Exception> processMessage(kj::ArrayPtr<const byte> message) {
+      // If `this` is the compressor, calling `processMessage()` will compress the `message`.
+      // Likewise, if `this` is the decompressor, `processMessage()` will decompress the `message`.
+      // If (de)compression is successful, the result is returned as a Vector, otherwise,
+      // an Exception is returned.
+
+      setInput(message.begin(), message.size());
+
+      kj::Vector<kj::byte> result;
+      KJ_IF_MAYBE(e, processLoop(result)) {
+        return kj::mv(*e);
+      }
+      return kj::mv(result);
+    }
+
+    kj::Maybe<kj::Exception> processLoop(kj::Vector<kj::byte>& output) {
+      // Since Zlib buffers the writes, we want to continue processing until there's nothing left.
+      Result result;
+      KJ_IF_MAYBE(exception, kj::runCatchingExceptions([this, &result]() {
+        result = pumpOnce();
+      })) {
+        return kj::mv(*exception);
+      }
+
+      if (result.buffer.size() > 0) {
+        output.addAll(result.buffer);
+      }
+
+      if (ctx.avail_in == 0 || result.processResult == Z_STREAM_END) {
```

Comment:
`avail_in` being zero doesn't necessarily mean we're done. It may be that zlib has consumed all of the input data into its internal state, but still has a few output bytes that didn't fit in the last buffer. We have to keep giving it new output buffers to write to until it actually signals to us that it has finished.

Reply:
I tried to rely on `Z_STREAM_END` only but we would end up in a scenario where we'd get `Z_BUF_ERROR` because the input was empty. Resolved it by checking for `Z_BUF_ERROR` showing up along with an empty output buffer. https://github.com/capnproto/capnproto/pull/1505/commits/6b4ff9d4b80d4af8a3cea510229e42fa260c2699

Reply:
Hmmm, that doesn't really sound right. Actually, I found this in `zlib.h`:

```
    If the parameter flush is set to Z_SYNC_FLUSH, all pending output is
  flushed to the output buffer and the output is aligned on a byte boundary, so
  that the decompressor can get all input data available so far.  (In
  particular avail_in is zero after the call if enough output space has been
  provided before the call.) Flushing may degrade compression for some
  compression algorithms and so it should be used only when necessary.  This
  completes the current deflate block and follows it with an empty stored block
  that is three bits plus filler bits to the next byte, followed by four bytes
  (00 00 ff ff).
```

This seems to say that checking `avail_in == 0` is actually sufficient. So maybe you were right initially. Sorry for the false alarm.

Reply:
I think this needs to be investigated a bit more. Earlier today I noticed [chromium](https://github.com/chromium/chromium/blob/c676681f0040f5b310f8949bd44d332b472a712d/net/websockets/websocket_deflater.cc#L73-L113) uses both `Z_NO_FLUSH` and `Z_SYNC_FLUSH`. I don't know for certain _yet_ [1], but I suspect that flushing to the output buffer when using `Z_NO_FLUSH` means zlib will not end the deflate block each time the 4KB buffer is full (whereas `Z_SYNC_FLUSH` would).

**If so**, I figure the drawback of using `Z_SYNC_FLUSH` is that for messages that are still large _after_ compression, we will have multiple deflate blocks (including empty blocks after each), whereas for `Z_NO_FLUSH`, the entire compressed message will fit in a single deflate block. Assuming a 1MB message limit (taking this from `receive()` in `http.h`) and a buffer of 4096 bytes, the largest uncompressed message would fit in about ~244 buffers. However, the compressed message would need to be smaller, so the upper bound on number of buffers (deflate blocks) is probably 200-220? That's likely less than 1KB of extra data going over the wire (worst case) per message.

Otherwise, it's not immediately obvious to me why we would want to use `Z_NO_FLUSH` in our case since:
1. We only start compressing/decompressing once the entire input (message) is available. If a message were to only _partially_ fill up the `output` buffer, it's fine, since we don't expect any more input (until we work on the next message, which is separate).
2. `Z_NO_FLUSH` seems to force a flush if the `output` buffer would be totally filled (`avail_out == 0`), so
i. In the case that we have a large message, we're forcing a flush either way.
ii. In the case that the message fits in the buffer, we end up calling `Z_SYNC_FLUSH` either way.

Anything obvious I'm missing here? I've messed around a bit with these two flush types a bit, but I still need to try sending some _really_ large messages to see if there's any significant difference.

[1]: `zlib.h` hasn't really explained flushing vs. finishing blocks, etc. I'll look again, but I think it might be easier to just check actual program output.

Reply:
Chromium uses both `Z_NO_FLUSH` and `Z_SYNC_FLUSH` because it doesn't necessarily write the whose message at once. Presumably `AddBytes()` can be called multiple times to add more data. If they flushed after every call to `AddBytes()`, then the compressor would be unable to take advantage of redundancies that cross a buffer boundary, because it is forced to completely process each buffer before it knows what comes in the next buffer.

In our case, though, we always present the _entire_ message to zlib at one time. So we can always pass `Z_SYNC_FLUSH`.

Note that if zlib stops because it doesn't have enough space in the output buffer, then it doesn't matter what flush mode you specified. The flush only actually applies if zlib was able to consume all input and write out everything it wanted to output.

To put it another way:
* You should use `Z_NO_FLUSH` to tell zlib: "This is not the whole message. There will be another buffer of input after this one. So you don't have to write out everything necessary to decode it yet since there's more data coming anyway."
* You use `Z_SYNC_FLUSH` to say "This is the last (or only) buffer in the message. Please make sure that you output all bytes needed so that the other end will be able to full decompress the message."

Reply:
Thanks for clarifying! https://github.com/capnproto/capnproto/pull/1505/commits/6a0e04cc8a99acec3fc99af92f8bfd9988e577cf